File: code/PID.c
#include "zf_common_headfile.h"

//Pos_PID Speed;
Pos_PID speed_left_pid = {0};
Pos_PID speed_right_pid = {0};
DIR_PID direction_pid = {0};
Increment_PID direction_pid_2 = {0};

/*速度环PID初始化*/
void Speed_PID_Init(void)
{

}

/*方向环PID初始化*/
void Direction_PID_Init(void)
{

}

/*PID最终初始化*/
void PID_Init(void)
{
    Speed_PID_Init();

    Direction_PID_Init();
}

/****************************PID计算核心函数************************************/

/*
 功能：位置式PID
 参数：pid - PID结构体指针 | expect - 目标值 | feedback - 反馈值 | dt - 时间间隔（单位：秒）
 返回值：计算得到的PID输出
*/
float Pos_PID_Calculate(Pos_PID *pid, float expect, float feedback, float dt)
{
    float Points_ratio = 0.0f;  // 积分权重系数，用于动态调整积分贡献

    // 计算当前误差（目标值 - 反馈值）
    pid->Error = expect - feedback;

    // 是否发车，防止未发车前积分累加
    if(1)
    {
        // 计算积分权重：误差越小，权重越大（非线性积分，减少大误差时的积分积累）
        // 公式：1 - (当前误差^2)/(最大误差^2)，限制在0~1之间
        Points_ratio = MAX((1-(pid->Error*pid->Error)/(pid->MAX_Error*pid->MAX_Error)),0);

        // 积分项累加（带权重，防止大误差时积分饱和）
        pid->Integral += Points_ratio * pid->Error * dt;

        // 积分限幅：限制积分项在[-MAX_Integral, MAX_Integral]范围内
        pid->Integral = LIMIT(pid->Integral, -pid->MAX_Integral, pid->MAX_Integral);
    }
    else
    {
        pid->Integral = 0.0f;
    }

    // 计算微分项：对反馈值的变化率（反馈值的导数）进行一阶低通滤波（LPF_1）
    // 作用：减少噪声对微分环节的影响
    LPF_1(10, dt, ((feedback - pid->Last_Feedback)/dt), &pid->Differential);

    // 计算PID输出：前馈项 + 比例项 + 积分项 - 微分项（微分作用于反馈变化率）
    pid->OutPut = pid->Kf * expect + pid->Kp * pid->Error + pid->Ki * pid->Integral - pid->Kd * pid->Differential;

    // 输出限幅：限制输出在[-MAX_OutPut, MAX_OutPut]范围内，保护执行器
    pid->OutPut = LIMIT(pid->OutPut, -pid->MAX_OutPut, pid->MAX_OutPut);

    // 保存当前反馈值，用于下一次计算微分
    pid->Last_Feedback = feedback;

    return pid->OutPut;
}

/*
 功能：增量式式PID
*/
float Increment_PID_Caculate(Increment_PID *pid, int16_t Target, int16_t feedback)
{
    pid->Error = Target - feedback;

    pid->Pout = pid->Kp*(pid->Error - pid->Last_Error);
    pid->Iout = pid->Ki*pid->Error;
    pid->Dout = pid->Kd*(pid->Error - 2*pid->Last_Error + pid->Prev_Error);

    pid->OutPut += pid->Pout + pid->Iout + pid->Dout;
    pid->OutPut = LIMIT(pid->OutPut, -pid->MAX_OutPut, pid->MAX_OutPut);    //限幅

    pid->Prev_Error = pid->Last_Error;
    pid->Last_Error = pid->Error;

    return pid->OutPut;
}

File: code/pid.py
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.widgets import Slider

# PID 参数
Kf = 0.6
Kp = 0.8
Ki = 0.1
Kd = 0.0
MAX_Error = 200.0
MAX_Integral = 500.0
MAX_OutPut = 5000.0
dt = 0.005

def limit(x, lower, upper):
    return max(min(x, upper), lower)

def pid_calculate(expect, feedback, integral):
    """
    支持负数 feedback/expect 的 PID 计算
    """
    error = expect - feedback
    # 允许负误差积分累加，不再强制 >=0
    points_ratio = 1 - (error**2)/(MAX_Error**2)
    points_ratio = limit(points_ratio, -1, 1)  # 限制在 [-1,1] 防爆积分
    integral += points_ratio * error * dt
    integral = limit(integral, -MAX_Integral, MAX_Integral)
    differential = 0
    OutPut = Kf*expect + Kp*error + Ki*integral - Kd*differential
    OutPut = limit(OutPut, -MAX_OutPut, MAX_OutPut)
    return OutPut, integral

# 初始化图形
expect_init = 1000
feedback_values = np.linspace(-8000, 8000, 500)
integral = 0.0
outputs = []
for fb in feedback_values:
    out, integral = pid_calculate(expect_init, fb, integral)
    outputs.append(out)

fig, ax = plt.subplots()
plt.subplots_adjust(bottom=0.25)
l, = plt.plot(feedback_values, outputs, lw=2, color='blue')
target_line = ax.axvline(expect_init, color='red', linestyle='--', label='Target value')

ax.set_xlabel('Feedback')
ax.set_ylabel('PID Output')
ax.set_title('PID Output vs Feedback')
ax.set_xlim(-8000, 8000)
ax.set_ylim(-5000, 5000)
ax.grid(True)
ax.legend()

# 添加滑块
ax_expect = plt.axes([0.2, 0.1, 0.65, 0.03])
slider_expect = Slider(ax_expect, 'Expect', 0, 5000, valinit=expect_init)

def update(val):
    expect = slider_expect.val
    feedback_values = np.linspace(-8000, 8000, 500)
    integral = 0.0
    outputs = []
    for fb in feedback_values:
        out, integral = pid_calculate(expect, fb, integral)
        outputs.append(out)
    l.set_xdata(feedback_values)
    l.set_ydata(outputs)
    target_line.set_xdata([expect, expect])
    # 坐标轴固定
    l.axes.set_xlim(-8000, 8000)
    l.axes.set_ylim(-5000, 5000)
    fig.canvas.draw_idle()

slider_expect.on_changed(update)
plt.show()
